core/src/main/java/org/jruby/compiler/BlockJITTask.java

```diff
@@ -34,71 +34,79 @@
 
-class BlockJITTask implements Runnable {
-    private JITCompiler jitCompiler;
+import static org.jruby.compiler.MethodJITTask.*;
+
+class BlockJITTask extends JITCompiler.Task {
+
     private final String className;
     private final MixedModeIRBlockBody body;
     private final String methodName;
 
     public BlockJITTask(JITCompiler jitCompiler, MixedModeIRBlockBody body, String className) {
-        this.jitCompiler = jitCompiler;
+        super(jitCompiler);
         this.body = body;
         this.className = className;
         this.methodName = body.getName();
     }
 
-    public void run() {
-        // We synchronize against the JITCompiler object so at most one code body will jit at once in a given runtime.
-        // This works around unsolved concurrency issues within the process of preparing and jitting the IR.
-        // See #4739 for a reproduction script that produced various errors without this.
-        synchronized (jitCompiler) {
-            try {
-                String key = SexpMaker.sha1(body.getIRScope());
-                JVMVisitor visitor = new JVMVisitor(jitCompiler.runtime);
-                BlockJITClassGenerator generator = new BlockJITClassGenerator(className, methodName, key, jitCompiler.runtime, body, visitor);
-
-                JVMVisitorMethodContext context = new JVMVisitorMethodContext();
-                generator.compile(context);
-
-                // FIXME: reinstate active bytecode size check
-                // At this point we still need to reinstate the bytecode size check, to ensure we're not loading code
-                // that's so big that JVMs won't even try to compile it. Removed the check because with the new IR JIT
-                // bytecode counts often include all nested scopes, even if they'd be different methods. We need a new
-                // mechanism of getting all body sizes.
-                Class sourceClass = visitor.defineFromBytecode(body.getIRScope(), generator.bytecode(), new OneShotClassLoader(jitCompiler.runtime.getJRubyClassLoader()));
-
-                if (sourceClass == null) {
-                    // class could not be found nor generated; give up on JIT and bail out
-                    jitCompiler.counts.failCount.incrementAndGet();
-                    return;
-                } else {
-                    generator.updateCounters(jitCompiler.counts, body.ensureInstrsReady());
+    @Override
+    public void exec() {
+        try {
+            // Check if the method has been explicitly excluded
+            String excludeModuleName = checkExcludedMethod(jitCompiler.config, className, methodName, body.getImplementationClass());
+            if (excludeModuleName != null) {
+                body.setCallCount(-1);
+                if (jitCompiler.config.isJitLogging()) {
+                    JITCompiler.log(body.getImplementationClass(), body.getFile(), body.getLine(), methodName, "skipping block in " + excludeModuleName);
```

Comment:
Why did this move in here? We can't really exclude blocks from JIT by method name, can we?

Reply:
believe we can - if a method is excluded its blocks should also not JIT. however I did not test if that case works - just noticed on aprod instance that blocks are being JITed for an excluded class. that case, an excluded mod or class, is reasonable and should be handled here ...